Orckestra.Widget.FilteredSelector/Orckestra.Widget.FilteredSelector/WidgetProvider/FilteredSelectorWidgetProcessingManager.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+
+using Composite.Core.Types;
+using Composite.Core.WebClient.Renderings.Page;
+using Composite.Data;
+
+namespace Orckestra.Widget.FilteredSelector.WidgetProvider
+{
+    internal static class FilteredSelectorWidgetProcessingManager
+    {
+        public static IEnumerable<Guid> GetAssociatedPageIds(Guid pageId, SitemapScope sitemapScope) => PageStructureInfo.GetAssociatedPageIds(pageId, sitemapScope);
+        private static IEnumerable<KeyValuePair> GetOptionsForDefault(Type t, Guid pageId, SitemapScope sitemapScope)
+        {
+            IEnumerable<Guid> availablePageIds = GetAssociatedPageIds(pageId, sitemapScope).Distinct();
+
+            using (DataConnection con = new DataConnection())
+            {
+                MethodInfo generic = typeof(DataConnection).GetMethod("Get").MakeGenericMethod(t);
+                IEnumerable<KeyValuePair> result = from a in (IEnumerable<IPageRelatedData>)generic.Invoke(con, null)

```

Comment:
it makes sense not to evaluate "a.GetKeyProperties().First(k => k.Name == "Id")" on every loop iteration.

```
var idProperty = a.GetKeyProperties().First(k => k.Name == "Id");
...
select new KeyValuePair(idProperty.GetValue(a).ToString(), a.GetLabel());
```

Reply:
Updated